src/Features/ExternalAccess/Copilot/Internal/Completion/CSharpContextProviderService.cs

```diff
@@ -6,80 +6,34 @@
+[method: ImportingConstructor]
+[method: Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+internal sealed class CSharpContextProviderService([ImportMany] IEnumerable<IContextProvider> providers)
+    : ICSharpCopilotContextProviderService
 {
-    // Exposed for testing
-    public ImmutableArray<IContextProvider> Providers { get; }
-
-    [ImportingConstructor]
-    [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
-    public CSharpContextProviderService([ImportMany] IEnumerable<IContextProvider> providers)
-    {
-        Providers = providers.ToImmutableArray();
-    }
-
-    public async IAsyncEnumerable<IContextItem> GetContextItemsAsync(Document document, int position, IReadOnlyDictionary<string, object> activeExperiments, [EnumeratorCancellation] CancellationToken cancellationToken)
-    {
-        var queue = new AsyncQueue<IContextItem>();
-        var tasks = this.Providers.Select(provider => Task.Run(async () =>
-        {
-            try
-            {
-                await provider.ProvideContextItemsAsync(document, position, activeExperiments, ProvideItemsAsync, cancellationToken).ConfigureAwait(false);
-            }
-            catch (Exception exception) when (FatalError.ReportAndCatchUnlessCanceled(exception, ErrorSeverity.General))
-            {
-            }
-        },
-        cancellationToken));
-
-        // Let all providers run in parallel in the background, so we can steam results as they come in.
-        // Complete the queue when all providers are done.
-        _ = Task.WhenAll(tasks)
-            .ContinueWith((_, __) => queue.Complete(),
-                          null,
-                          cancellationToken,
-                          TaskContinuationOptions.ExecuteSynchronously,
-                          TaskScheduler.Default);
-
-        while (true)
-        {
-            IContextItem item;
-            try
-            {
-                item = await queue.DequeueAsync(cancellationToken).ConfigureAwait(false);
-            }
-            catch (OperationCanceledException) when (!cancellationToken.IsCancellationRequested)
-            {
-                // Dequeue is cancelled because the queue is empty and completed, we can break out of the loop.
-                break;
-            }
-
-            yield return item;
-        }
-
-        ValueTask ProvideItemsAsync(ImmutableArray<IContextItem> items, CancellationToken cancellationToken)
-        {
-            foreach (var item in items)
-            {
-                queue.Enqueue(item);
-            }
-
-            return default;
-        }
-    }
+    private readonly ImmutableArray<IContextProvider> _providers = [.. providers];
+
+    public IAsyncEnumerable<IContextItem> GetContextItemsAsync(Document document, int position, IReadOnlyDictionary<string, object> activeExperiments, CancellationToken cancellationToken)
+        => ProducerConsumer<IContextItem>.RunParallelStreamAsync(
```

Comment:
this automatically parallelizes, prefering to utilize all cores to finish a fixed set of work that is then added to as work items complete. this is better than kicking off N items (which may be far mroe than the number of cores), causing them to contend with each other, and trying to .WhenAll them.

Reply:
thanks! I initially tried to use roslyn's ProducerConsumer but it quickly ended up with the problem of requiring too much code to be exposed via EA, so I wrote my own parallelization instead. Now that we moved this code over, it makes perfect sense to fix it